lighthouse-core/test/audits/seo/crawlable-anchors-test.js

```diff
@@ -0,0 +1,143 @@
+    assert.equal(runAudit({rawHref: '#:~:text=string'}), 1, 'hyperlink with a text fragment');
+    assert.equal(runAudit({rawHref: 'ftp://myname@host.dom'}), 1, 'an FTP hyperlink');
+    assert.equal(runAudit({rawHref: 'http://172.217.20.78'}), 1, 'IP address based link');
+    assert.equal(runAudit({rawHref: '//example.com'}), 1, 'protocol relative link');
+    assert.equal(runAudit({rawHref: 'tel:5555555'}), 1, 'email link with a tel URI');
+    assert.equal(runAudit({rawHref: '#'}), 1, 'link with only a hash symbol');
+    assert.equal(runAudit({
+      rawHref: '?query=string',
+    }), 1, 'relative link which specifies a query string');
+  });
+
+  it('allows anchors which use a name attribute', () => {
+    assert.equal(runAudit({name: 'name'}), 1, 'link with a name attribute');
+  });
+
+  it('handles anchor elements which use event listeners', () => {
+    const auditResultClickPresent = runAudit({
+      listeners: [{type: 'click'}],
+    });
+    assert.equal(auditResultClickPresent, 1, 'presence of a click handler is a pass');
+
+    const auditResultJavaScriptURI = runAudit({
+      rawHref: 'javascript:void(0)',
+      listeners: [{type: 'click'}],
+    });
+    const assertionMessage = 'hyperlink with a `javascript:` URI and a click handler';
+    assert.equal(auditResultJavaScriptURI, 1, assertionMessage);
+
+    const auditResultNonClickListener = runAudit({
+      listeners: [{type: 'nope'}],
+    });
+    assert.equal(auditResultNonClickListener, 0, 'no click event is a fail');
+
+    const auditResultMixtureOfListeners = runAudit({
+      listeners: [{type: 'nope'}, {type: 'another'}, {type: 'click'}],
+    });
+    assert.equal(auditResultMixtureOfListeners, 1, 'at least one click listener is a pass');
+  });
+
+  it('disallows uncrawlable anchors', () => {
+    assert.equal(runAudit({}), 0, 'link with no meaningful attributes and no event handlers');
+    assert.equal(runAudit({rawHref: 'file:///image.png'}), 0, 'hyperlink with a `file:` URI');
+    assert.equal(runAudit({name: ' '}), 0, 'name attribute with only space characters');
+    assert.equal(runAudit({rawHref: ' '}), 0, 'href attribute with only space characters');
+    const assertionMessage = 'onclick attribute with only space characters';
+    assert.equal(runAudit({rawHref: ' ', onclick: ' '}), 0, assertionMessage);
+  });
+
+  it('handles javascript:void expressions in the onclick attribute', () => {
+    const expectedAuditFailures = [
+      'javascript:void(0)',
+      'javascript: void(0)',
+      'javascript : void(0)',
+      'javascript : void ( 0 )',
+      'javascript: void 0',
+      'javascript:void 0',
+      // The audit logic removes all whitespace from the string and considers this a fail
+      'javascript:void0',
+    ];
+
+    for (const javaScriptVoidVariation of expectedAuditFailures) {
+      const auditResult = runAudit({rawHref: javaScriptVoidVariation});
+      assert.equal(auditResult, 0, 'javascript:void failing variations');
```

Comment:
for these loop ones it'd be nice to stick the `variation` into the failing message for easier debugging when it fails

Reply:
Good catch